File: samples/SampleQueryableSqlite/Provider/ProductDemo.cs
using System.Linq;
using Metano.Annotations;

namespace SampleQueryableSqlite.Provider;

/// <summary>
/// User-facing entry points for the sample. Each method composes a
/// LINQ chain over <see cref="IProductRepository.Products"/> and
/// hands the resulting <see cref="IQueryable{Product}"/> to the
/// repository's terminal method.
///
/// <para>
/// The chain itself is captured by Phase B (#196) — every lambda
/// body is encoded as an <c>ExprTree</c> alongside the closure. The
/// runtime's <c>getStages</c> helper (#200) makes that descriptor
/// list visible on the materialized result, so the TS-side adapter
/// can translate the trees into SQL and execute them via
/// <c>bun:sqlite</c> instead of evaluating the closures.
/// </para>
/// </summary>
[Transpile]
public static class ProductDemo
{
    public static Product[] ActiveProducts(IProductRepository repo) =>
        repo.ToArray(repo.Products.Where(p => p.IsActive));

    public static Product[] InStock(IProductRepository repo) =>
        repo.ToArray(repo.Products.Where(p => p.IsActive && p.StockCount > 0));

    public static Product[] AtLeastPrice(IProductRepository repo, decimal minPrice) =>
        repo.ToArray(repo.Products.Where(p => p.UnitPrice >= minPrice));

    public static Product[] TopByPrice(IProductRepository repo, int topN) =>
        repo.ToArray(repo.Products.OrderByDescending(p => p.UnitPrice).Take(topN));

    public static Product[] Page(IProductRepository repo, int offset, int pageSize) =>
        repo.ToArray(repo.Products.OrderBy(p => p.Id).Skip(offset).Take(pageSize));

    public static int CountActive(IProductRepository repo) =>
        repo.Count(repo.Products.Where(p => p.IsActive));

    public static Product? FirstById(IProductRepository repo) =>
        repo.FirstOrDefault(repo.Products.OrderBy(p => p.Id));
}

File: targets/js/sample-queryable-sqlite/src/provider/product-demo.ts
/** biome-ignore-all lint/complexity/noUselessConstructor: explicit shape preserved by transpiler */
import { linq, orderBy, orderByDescending, skip, take, where } from "metano-runtime/system/linq";
import type { Decimal } from "decimal.js";
import type { Product } from "#";
import type { IProductRepository } from "./i-product-repository";

export class ProductDemo {
  constructor() {}

  static activeProducts(repo: IProductRepository): Product[] {
    return repo.toArray(
      linq(
        repo.products,
        where((p: Product) => p.isActive, {
          tree: {
            kind: "member",
            target: {
              kind: "param",
              name: "p",
              type: { name: "Product" },
            },
            member: "isActive",
          },
        }),
      ),
    );
  }

  static inStock(repo: IProductRepository): Product[] {
    return repo.toArray(
      linq(
        repo.products,
        where((p: Product) => p.isActive && p.stockCount > 0, {
          tree: {
            kind: "binary",
            op: "&&",
            left: {
              kind: "member",
              target: {
                kind: "param",
                name: "p",
                type: { name: "Product" },
              },
              member: "isActive",
            },
            right: {
              kind: "binary",
              op: ">",
              left: {
                kind: "member",
                target: {
                  kind: "param",
                  name: "p",
                  type: { name: "Product" },
                },
                member: "stockCount",
              },
              right: {
                kind: "literal",
                value: 0,
                type: { name: "number" },
              },
            },
          },
        }),
      ),
    );
  }

  static atLeastPrice(repo: IProductRepository, minPrice: Decimal): Product[] {
    return repo.toArray(
      linq(
        repo.products,
        where((p: Product) => p.unitPrice.gte(minPrice), {
          tree: {
            kind: "binary",
            op: ">=",
            left: {
              kind: "member",
              target: {
                kind: "param",
                name: "p",
                type: { name: "Product" },
              },
              member: "unitPrice",
            },
            right: {
              kind: "capture",
              name: "minPrice",
              type: { name: "number" },
            },
          },
          captures: { minPrice: minPrice },
        }),
      ),
    );
  }

  static topByPrice(repo: IProductRepository, topN: number): Product[] {
    return repo.toArray(
      linq(
        repo.products,
        orderByDescending((p: Product) => p.unitPrice, {
          tree: {
            kind: "member",
            target: {
              kind: "param",
              name: "p",
              type: { name: "Product" },
            },
            member: "unitPrice",
          },
        }),
        take(topN),
      ),
    );
  }

  static page(repo: IProductRepository, offset: number, pageSize: number): Product[] {
    return repo.toArray(
      linq(
        repo.products,
        orderBy((p: Product) => p.id, {
          tree: {
            kind: "member",
            target: {
              kind: "param",
              name: "p",
              type: { name: "Product" },
            },
            member: "id",
          },
        }),
        skip(offset),
        take(pageSize),
      ),
    );
  }

  static countActive(repo: IProductRepository): number {
    return repo.count(
      linq(
        repo.products,
        where((p: Product) => p.isActive, {
          tree: {
            kind: "member",
            target: {
              kind: "param",
              name: "p",
              type: { name: "Product" },
            },
            member: "isActive",
          },
        }),
      ),
    );
  }

  static firstById(repo: IProductRepository): Product | null {
    return repo.firstOrDefault(
      linq(
        repo.products,
        orderBy((p: Product) => p.id, {
          tree: {
            kind: "member",
            target: {
              kind: "param",
              name: "p",
              type: { name: "Product" },
            },
            member: "id",
          },
        }),
      ),
    );
  }
}
